### PR TITLE
hnerv_fec6_fixed_huffman_k16

### PR DESCRIPTION
**Pull request by adpena**
# submission name:

hnerv_fec6_fixed_huffman_k16

# upload zipped `archive.zip`

[archive.zip](https://github.com/adpena/comma_video_compression_challenge/releases/download/fec6-frontier-submission-20260520/archive.zip)
Runtime tree at PR head [`f2843e2`](https://github.com/adpena/comma_video_compression_challenge/tree/f2843e2b901543edd3afbce844a3fd71d83af7f7/submissions/hnerv_fec6_fixed_huffman_k16).

# report.txt

```
=== Evaluation config ===
  batch_size: 16
  device: cpu
  num_threads: 2
  prefetch_queue_depth: 4
  report: /root/modal_auth_eval_cpu_work/eval_work/report.txt
  seed: 1234
  submission_dir: /root/modal_auth_eval_cpu_work/eval_work
  uncompressed_dir: /workspace/pact/upstream/videos
  video_names_file: /workspace/pact/upstream/public_test_video_names.txt
=== Evaluation results over 600 samples ===
  Average PoseNet Distortion: 0.00002943
  Average SegNet Distortion: 0.00056029
  Submission file size: 178,517 bytes
  Original uncompressed size: 37,545,489 bytes
  Compression Rate: 0.00475469
  Final score: 100*segnet_dist + sqrt(10*posenet_dist) + 25*rate = 0.19
=== Archive identity ===
  Archive SHA-256: 6bae0201fb082457a02c69565531aba4c5942669c384fdc48e7d554f7b893fcf
  Archive size bytes: 178517
```

# does your submission require gpu for evaluation (inflation)?

no

# did you include the compression script? and want it to be merged?

yes. `compress.sh` wraps the encoder bundled at `submissions/hnerv_fec6_fixed_huffman_k16/encoder/`. The encoder reproduces this submission's `archive.zip` from PR [#101](https://github.com/commaai/comma_video_compression_challenge/pull/101)'s `archive.zip` (SHA-256 `b83bf3488625dbd73adeddff91712994197ab53098e578e91327a0c6e49efb3e`) plus the precomputed offline scorer-sweep artifacts used by the exact recipe, including the final 16-pair CPU overlay. PR #101's archive is not redistributed; fetch from its release. Full recipe in `encoder/README.md`.

# is this submission competitive or innovative? explain why

**Competitive: yes.** `0.192051 [contest-CPU]` vs PR [#101](https://github.com/commaai/comma_video_compression_challenge/pull/101) `hnerv_ft_microcodec` (`0.192840`, #1 prize per @YassineYousfi's comment), delta `-0.000789`. The selector adds +259 bytes (rate `+25 × 259 / 37,545,489 = +0.000172`) and saves `-0.000961` across the distortion terms; distortion savings are **~5.6× the rate cost**.

**Innovative: yes.** The submission stacks a per-pair perturbation selector on top of PR #101's frozen substrate, picked offline against the upstream SegNet+PoseNet response. The offline sweep tool (`encoder/frame_exploit_segnet_posenet_sweep.py`, macOS CPU local) scores 31 candidate per-frame perturbations (integer luma/chroma biases, single-pixel rolls, 8×8 signed chroma tiles, applied to either frame of the pair) against both scorers. The encoder (`encoder/build_pr101_frame_exploit_selector_packet.py`) picks K=16 modes from that table, and for each of the 600 frame-pairs ships the per-pair winning index over a fixed Huffman codebook (~3.24 bits/pair empirical). At inflate time, inflate.py reads each pair's index and `src/frame_selector.py` applies the corresponding K=16-mode transform to that pair's target frame (each mode carries its own `(transform, target_frame)` pair; the search is compress-time only and no scorer access happens at inflate). The scorer is asymmetric across pair positions: SegNet reads only the second frame, PoseNet reads both. So frame-0 perturbations move PoseNet alone, frame-1 perturbations move both scorers, and no single mode dominates across all 600 pairs. For this archive the K=16 winners are all frame-0; frame-1 candidates were swept but did not earn slots. Frame-0 is structurally invisible to SegNet, so frame-0 perturbations pay zero SegNet cost; the submission's name refers to this asymmetry. PR #101's weights are reused byte-identically; only the per-pair selector indices are transmitted. The sweep + selector pipeline applies to other frozen-substrate PRs.

# additional comments

FEC6 selector bolt-on around the public HNeRV lineage. Decoder (`src/model.py`) byte-identical to PR [#95](https://github.com/commaai/comma_video_compression_challenge/pull/95) (@AaronLeslie138); HNeRV architecture from [Chen et al. 2023](https://arxiv.org/abs/2304.02633) ([haochen-rye/HNeRV](https://github.com/haochen-rye/HNeRV)). Byte substrate from PR [#101](https://github.com/commaai/comma_video_compression_challenge/pull/101) (@SajayR): compact decoder, latent-payload parsing, Brotli source streams ([RFC 7932](https://www.rfc-editor.org/rfc/rfc7932)), canonical Huffman on the latent sidecar (all reused unchanged). Member `x` grammar: `FP11 | u32 source_len | source_pr101_payload | u16 selector_len | selector_payload`. PR #101's payload reads verbatim from `source_pr101_payload`; the FEC6 selector is byte-appended outside the Brotli envelope and is not further compressed. Lineage: PR [#98](https://github.com/commaai/comma_video_compression_challenge/pull/98) (@EthanYangTW), PR [#100](https://github.com/commaai/comma_video_compression_challenge/pull/100) (@BradyMeighan), PR [#102](https://github.com/commaai/comma_video_compression_challenge/pull/102) (@EthanYangTW), PR [#103](https://github.com/commaai/comma_video_compression_challenge/pull/103) (@rem2). Paired Modal Tesla T4 against the same `archive.zip` bytes returned `0.226210 [contest-CUDA]`; CPU and CUDA scored separately on the same archive bytes. Decoded RGB frames are bit-stable per device but differ across CPU and CUDA because `F.interpolate(..., mode='bicubic')` and the `clamp/round/uint8` cast are not bit-identical across the two backends. Bundles the compression script and standard submission housekeeping; full listing at the runtime-tree link above. Research notes at [`adpena/comma-lab`](https://github.com/adpena/comma-lab); tooling at [`adpena/tac`](https://github.com/adpena/tac).

**Further improvement** could come from widening the frame-0 perturbation catalog (frame-0 is the free axis against SegNet, so exhaustive search there is cheap), and from developing a distinct frame-1 methodology since perturbations there move both scorers; per-pair scorer-gradient signal at the SegNet class-region level could inform which perturbation classes apply where. The global K=16 catalog is the structural bottleneck: splitting it into per-tier or per-region menus (one menu for the hardest pairs by master-gradient norm, another for easier pairs) lets each menu optimize for its own pair population, and the encoder's pair-count-maximizing K=16 selection criterion could be replaced with per-pair-EV-weighted selection favoring high-impact-low-coverage modes. The catalog could also expand to both-frame composite modes (transforms targeting frame-0 and frame-1 simultaneously); the scorer is non-linear so composite Δ can differ from sum of individual Δs. Open directions include: a PoseNet-null subset of the frame-0 catalog (per-pair PoseNet-gradient bottom-decile → pure free-rate against both scorer axes); a SegNet-class-region waterfill following Fridrich UNIWARD's distortion-detector-aware embedding (per-pixel logit-margin as sensitivity prior); time-coherent run-length encoding exploiting the empirical conditional-entropy gap on the selector stream (H(next|prev) = 2.94 bits/pair vs marginal H = 3.21 ≈ 22 bytes recoverable); and per-pair counterfactual probes generated via gradient ascent against scorer disagreement.

**Quick reproducibility check (≈60 s, CPU only):**
```bash
curl -L -o $TMPDIR/archive.zip https://github.com/adpena/comma_video_compression_challenge/releases/download/fec6-frontier-submission-20260520/archive.zip
shasum -a 256 $TMPDIR/archive.zip                # expect 6bae0201fb082457a02c69565531aba4c5942669c384fdc48e7d554f7b893fcf
mkdir -p $TMPDIR/data $TMPDIR/out && unzip -oq $TMPDIR/archive.zip -d $TMPDIR/data && echo "0.mkv" > $TMPDIR/list.txt
PACT_INFLATE_DEVICE=cpu ./inflate.sh $TMPDIR/data $TMPDIR/out $TMPDIR/list.txt && shasum -a 256 $TMPDIR/out/0.raw  # expect d1afc583b01ff4a7aaa844d4f03ece3ed381d56763a06cb2c5e011526e5f868c
```

### CODE REVIEW

**Comment by YassineYousfi** [comment]
can you update the pr with the new template, including an easy to understand response to: 
```
# is this submission competitive or innovative? explain why
<!-- competitive: better than top #1 submission -->
<!-- innovative: it has a novel idea that is not on the leaderboard yet, might not be competitive, but has potential -->

```

**Comment by adpena** [comment]
I've made those changes, also bundled the encoder and sweep scripts.

**Comment by adpena** [comment]
Follow-up: I made the bundled encoder/sweep scripts self-contained, removed private/default experiment paths, added an explicit --upstream scorer-root input for the sweep, corrected the exact CPU-overlay reproduction recipe, and refreshed the runtime-tree link to the latest PR head. Release archive remains unchanged at SHA-256 6bae0201fb082457a02c69565531aba4c5942669c384fdc48e7d554f7b893fcf.
